File: Python/111.Minimum Depth of Binary Tree.py
#求叶结点的最小深度
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def inorderTraversal(self, root):
        self.mindepth , self.init = 0 , 0
        self.inorder(root,1)
        return self.mindepth

    def inorder(self,p,depth):
        if p!=None:
            if p.left==None and p.right==None:
                if self.init==0:
                    self.mindepth = depth
                    self.init = 1
                else:
                    self.mindepth = depth if self.mindepth>depth else self.mindepth
            else:
                self.inorder(p.left,depth+1) if p.left!=None else None
                self.inorder(p.right,depth+1) if p.right!=None else None


t1 = TreeNode(1)
t2 = TreeNode(2)
t3 = TreeNode(3)
t4 = TreeNode(4)
t5 = TreeNode(5)
t6 = TreeNode(6)
t1.left = t2
t1.right = t3
t3.left = t4
t4.right = t5

s = Solution()
print(s.inorderTraversal(t7))

File: C++/Tree/111.Minimum Depth of Binary Tree.cpp
/*
求二叉树最小深度，也就是从根到叶的最短路径长度。
*/
/*
思路：递归解法。左子树为空，递归右侧。右子树为空，递归左侧。否则两侧都递归取最小的。
*/
#include <iostream>
#include <windows.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
public:
	int minDepth(TreeNode* root) {
        if(!root)	return 0;
        if(!root->left) return minDepth(root->right)+1;
        if(!root->right) return minDepth(root->left)+1;
        return min(minDepth(root->left),minDepth(root->right))+1;
    }
};
int main()
{
	Solution s;
	TreeNode root(1),t1(2),t2(3);
	root.right = &t1;
	t1.left = &t2;
	cout << s.minDepth(&root)<<endl;
}

